sdks/python/apache_beam/runners/worker/worker_pool_main.py

```diff
@@ -0,0 +1,170 @@
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""
+Worker pool entry point.
+
+The worker pool exposes an RPC service that is used with EXTERNAL
+environment to start and stop the SDK workers.
+
+This entry point is used by the Python SDK container in worker pool mode.
+"""
+
+from __future__ import absolute_import
+
+import argparse
+import atexit
+import logging
+import subprocess
+import sys
+import threading
+import time
+from concurrent import futures
+
+import grpc
+
+from apache_beam.portability.api import beam_fn_api_pb2
+from apache_beam.portability.api import beam_fn_api_pb2_grpc
+from apache_beam.runners.worker import sdk_worker
+
+
+class BeamFnExternalWorkerPoolServicer(
+    beam_fn_api_pb2_grpc.BeamFnExternalWorkerPoolServicer):
+
+  def __init__(self, worker_threads, use_process=False,
+               container_executable=None):
+    self._worker_threads = worker_threads
+    self._use_process = use_process
+    self._container_executable = container_executable
+    self._worker_processes = {}
+
+  @classmethod
+  def start(cls, worker_threads=1, use_process=False, port=0,
+            container_executable=None):
+    worker_server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
+    worker_address = 'localhost:%s' % worker_server.add_insecure_port(
+        '[::]:%s' % port)
+    beam_fn_api_pb2_grpc.add_BeamFnExternalWorkerPoolServicer_to_server(
+        cls(worker_threads,
+            use_process=use_process,
+            container_executable=container_executable),
+        worker_server)
+    worker_server.start()
+
+    # Register to kill the subprocesses on exit.
+    def kill_worker_processes():
+      for worker_process in cls._worker_processes.values():
+        worker_process.kill()
+    atexit.register(kill_worker_processes)
+
+    return worker_address, worker_server
+
+  def StartWorker(self, start_worker_request, unused_context):
+    try:
+      if self._use_process:
+        command = ['python', '-c',
+                   'from apache_beam.runners.worker.sdk_worker '
+                   'import SdkHarness; '
+                   'SdkHarness("%s",worker_count=%d,worker_id="%s").run()' % (
+                       start_worker_request.control_endpoint.url,
+                       self._worker_threads,
+                       start_worker_request.worker_id)]
+        if self._container_executable:
+          # command as per container spec
+          command = [self._container_executable,
+                     '--id=%s' % start_worker_request.worker_id,
+                     '--logging_endpoint=%s'
+                     % start_worker_request.logging_endpoint.url,
```

Comment:
Does endpoint url's need to be part of the flags here? Presumably we can read these from the environment and pass it back again to avoid any inconsistency with the initial endpoint urls used for at the container startup.

Reply:
Endpoint URLs are not known at container startup. That's also why artifact staging has to happen lazily. I will document this.

Reply:
Example how the worker pool is started: `docker run --rm -p=50000:50000 tweise-docker-apache.bintray.io/beam/python:latest --worker_pool=true`

Reply:
Makes sense. Thank you for dcoumenting.